### PR TITLE
feat(theming): add focus styling utilities

### PR DESCRIPTION
**Pull request by jzempel**
## Description

This PR adds `xs` units to the `DEFAULT_THEME` to account for updated "spacer" focus styling, along with three new utilities:

- `focusStyles` (preferred) provides a CSS block with focus `outline` reset, focus-visible `box-shadow` styling, and [transparent outline hack](https://css-tricks.com/quick-tips-for-high-contrast-mode/) for high contrast mode
- `getFocusBoxShadow` (fallback) in the case where existing code would be a tedious/surgical refactor, this utility provides the CSS value for the `box-shadow` property
- `SELECTOR_FOCUS_VISIBLE` constant that can be used together with the `selector` parameter on `focusStyles`:

```js
focusStyles({ theme: props.theme, selector: `&:focus-within, ${SELECTOR_FOCUS_VISIBLE}` });
```

## Detail

For performance concerns, the `getColor` utility was memoized, resulting in an extensive diff due to the `memoize` function wrap.

This PR contains two commits (rolled back), demonstrating real utility usage within a couple of key styled components:
- [StyledTextInput](https://github.com/zendeskgarden/react-components/pull/1542/commits/d1ea6a352faf989f4e98a11aba4d778fc5e4fa5f) using `focusStyles` (note that the existing `:focus` block should've also been removed)
- [StyledButton](https://github.com/zendeskgarden/react-components/pull/1542/commits/290b4f6e1c08cc377515f58eb302ea78bb3e1454) using `getFocusBoxShadow` (`focusStyles` could be applied with extra effort)

## Checklist

- [x] :ok_hand: design updates will be Garden Designer approved (add the designer as a reviewer)
- [ ] :globe_with_meridians: ~demo is up-to-date (`yarn start`)~
- [ ] :arrow_left: ~renders as expected with reversed (RTL) direction~
- [x] :metal: renders as expected with Bedrock CSS (`?bedrock`)
- [x] :guardsman: includes new unit tests. Maintain [existing coverage](https://coveralls.io/github/zendeskgarden/react-components?branch=main) (always >= 96%)
- [x] :wheelchair: tested for [WCAG 2.1 AA](https://www.w3.org/WAI/WCAG21/quickref/?currentsidebar=%23col_customize&levels=aaa) accessibility compliance
- [x] :memo: tested in Chrome, Firefox, Safari, and Edge

### CODE REVIEW

File: packages/theming/src/utils/getColor.ts [hunk 1]
```diff
@@ -37,57 +38,53 @@ const adjust = (color: string, expected: number, actual: number) => {
...
-) {
-  let retVal;
-
-  if (isNaN(shade)) {
-    return undefined;
-  }
-
-  const palette = theme && theme.palette ? theme.palette : DEFAULT_THEME.palette;
-  const colors = theme && theme.colors ? theme.colors : DEFAULT_THEME.colors;
+export const getColor = memoize(
```
Comment (Francois-Esquire) [inline packages/theming/src/utils/getColor.ts:42]: Were there any performance concerns that prompted the `lodash.memoize` utility usage?

File: packages/theming/src/utils/focusStyles.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ *   box-shadow: 0 0 0 {1px} #fff,
+ *               0 0 0 {3px} {blue};
+ *   outline: {2px} solid transparent;
+ *   outline-offset: {1px};
+ *   // additional {styles} here...
+ * }
+ * ```
+ */
+export const focusStyles = ({
+  condition = true,
```
Comment (Francois-Esquire) [inline packages/theming/src/utils/focusStyles.ts:46]: I'm curious, what advantage does the `condition` argument bring? The `condition` argument feels forced and can be done in the context of the code where needed (which reads clearer and less surface area to debug).

File: packages/theming/src/utils/focusStyles.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  const outline = `${math(
+    `${theme.shadowWidths[shadowWidth]} - ${theme.shadowWidths[spacerWidth]}`
+  )} solid transparent`;
+  const boxShadow = condition
+    ? getFocusBoxShadow({ shadowWidth, spacerWidth, theme, ...options })
+    : undefined;
+
+  /*
+   * 1. Browser reset
+   * 2. High contrast mode hack
```
Comment (Francois-Esquire) [inline packages/theming/src/utils/focusStyles.ts:62]: - Can you please share your resource/research on the high contrast mode hack? Very curious to learn more about this approach.
- Will there need to be other parts to this mechanism? Maybe an extension to the theme colors in the future to set the outline color w/ query?

File: packages/theming/src/utils/focusStyles.ts (hunk 2 above)
Comment (jzempel) [reply packages/theming/src/utils/focusStyles.ts:46]: See [commit](https://github.com/zendeskgarden/react-components/pull/1542/commits/d1ea6a352faf989f4e98a11aba4d778fc5e4fa5f) linked in description

File: packages/theming/src/utils/focusStyles.ts (hunk 3 above)
Comment (jzempel) [reply packages/theming/src/utils/focusStyles.ts:62]: See [link](https://css-tricks.com/quick-tips-for-high-contrast-mode/) in description

File: packages/theming/src/utils/getColor.ts (hunk 1 above)
Comment (jzempel) [reply packages/theming/src/utils/getColor.ts:42]: Yes, calling out to these utilities results in repeat `getColor` calls in a lot of existing cases. But I think there's a massive savings potential overall.

**Review by geotrev — COMMENTED** [review]
Nothing major to comment on. I dig it.

**Review by geotrev — APPROVED** [review]

File: packages/theming/src/utils/getColor.ts [hunk 4]
```diff
@@ -37,57 +38,53 @@ const adjust = (color: string, expected: number, actual: number) => {
...
-  }
+    if (transparency) {
+      retVal = rgba(retVal, transparency);
+    }
 
-  return retVal;
-}
+    return retVal;
+  },
+  (...args) => JSON.stringify(args)
```
[the remark this answers is not in the document]
Comment (jzempel) [reply packages/theming/src/utils/getColor.ts:89]: Yeah, I'm also concerned about this resolver as it pertains to the `theme` object. I pushed in on this pretty hard, and profiling seems to indicate the performance is plenty fast. I'm of two minds here 1) rollback the update, 2) push the update and be prepared to rollback if there are unintended performance consequences. I'm leaning toward 2.

File: packages/theming/src/utils/focusStyles.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import { css, CSSObject } from 'styled-components';
+import { math } from 'polished';
+import { FocusBoxShadowParameters, getFocusBoxShadow } from './getFocusBoxShadow';
+
+export const SELECTOR_FOCUS_VISIBLE = '&:focus-visible, &[data-garden-focus-visible="true"]';
```
Comment (geotrev) [inline packages/theming/src/utils/focusStyles.ts:12]: `:focus-visible` 🚀 🚀

File: packages/theming/src/utils/getColor.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (geotrev) [reply packages/theming/src/utils/getColor.ts:89]: Seems fair to me. Knowing the theme object is pretty shallow also mitigates my deepest concern anyway (no pun intended).

**Review by Francois-Esquire — APPROVED** [review]
Looks sharp!

**Pull request merged**